Add configuration local app data support

NapOli1084 opened the pull request:
The SearchConfiguration.xml file and search indexes were saved directly
at the location of the executing assembly, which meant that different
configuration and indexes were used when we ran the GUI Viewer vs the
CLI application or when we ran Debug vs Release. It also meant data
files were saved along with the binaries.

The new behavior first checks if there is a configuration file at the
location of the executing assembly and uses it if that's the case. That
allows backward compatibility for users that have their configuration at
that location. However, if there is no configuration at that location,
we now look into `%LocalAppData%/FlashSearcher` and save the default
configuration there if there's isn't one already. That allows sharing a
single configuration between GUI/CLI and Debug/Release, and saves user
data in her/his own folder.

NapOli1084 committed:
Add configuration local app data support

The SearchConfiguration.xml file and search indexes were saved directly
at the location of the executing assembly, which meant that different
configuration and indexes were used when we ran the GUI Viewer vs the
CLI application or when we ran Debug vs Release. It also meant data
files were saved along with the binaries.

The new behavior first checks if there is a configuration file at the
location of the executing assembly and uses it if that's the case. That
allows backward compatibility for users that have their configuration at
that location. However, if there is no configuration at that location,
we now look into `%LocalAppData%/FlashSearcher` and save the default
configuration there if there's isn't one already. That allows sharing a
single configuration between GUI/CLI and Debug/Release, and saves user
data in her/his own folder.

pascaloon commented on FlashSearch.Configuration/ConfigurationPathResolver.cs:
```diff
@@ -0,0 +1,71 @@
+        {
+            string directoryName = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+            if (String.IsNullOrEmpty(directoryName))
+                return ConfigFileName;
```
Why not throw?

NapOli1084 replied:
Because I didn't want that to prevent looking into app data folder. But I guess not finding the assembly path is not very likely anyway. If you prefer I can put back the throw.

pascaloon replied:
The only way I couldn't think where the location of an assembly is null is when it's loaded as a byte array.
I'd prefer a throw or return null.

NapOli1084 replied:
Ok, will do!

pascaloon commented on FlashSearch.Viewer/ViewModels/SearchViewModel.cs:
```diff
@@ -242,8 +242,7 @@ private void Search()
             
-            string localDirectory = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
-            string indexDirectory = Path.Combine(localDirectory, "Indexes", SelectedProject.Name, SelectedFileFilter.Index);
+            string indexDirectory = ConfigurationPathResolver.GetIndexDir(SelectedProject.Name, SelectedFileFilter.Index);
```
Could you refactor `ConfigurationPathResolver` into an injectable service? (like for the file service for example)
I'm not too fond of Singletons

NapOli1084 replied:
Ok, I'm not too familiar with injectable services, but I'll look at the file service.

NapOli1084 committed:
Refactor `ConfigurationPathResolver` as an injected service

This change refactors the `ConfigurationPathResolver` class to be used
as an injected service rather than a singleton.

NapOli1084 committed:
Throw when unable to find executable directory

pascaloon commented:
Alright! Thanks for the MR ❤️ !
